### PR TITLE
Added BYOL loss to losses and simple MLP to models

### PR DESCRIPTION
**Pull request by laurenmoos**
**this does NOT have** 
1. model refactoring (as I will iteratively follow your changes once you cement the parent classes)
2. momentum (it is passed as parameter but the logic isn't implemented yet)

**this does have:**
1. models (simple MLP)  under models
2. loss (forward pass) under loss
3. I have way too many classes in the loss file but perhaps this will give you/us inspiration for some core abstractions that would be helpful (wrapper class for models to access hidden states/dual-model & consistency loss abstractions & augmentation abstractions)- BYOL is simple logic but somewhat complicated framework so needs a lot of OOP

**my next steps would be the following** 
1. adding a benchmark/testing (I think it would be cool to demonstrate use of batch norm  influencing functionality!)  and training loop
2. adding momentum implementation
3. waiting for your commit to make architecture homogeneous with the higher level abstractions you propose (probably should have abstractions of Online/Target model as well as the low level model architectures

### CODE REVIEW

File: lightly/loss/byol_loss.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
+import singleton as singleton
+import torch
+from torch import nn
+
+import torch.nn.functional as F
+
+from models.mlp import MLP
+
+
```
Comment (philippmwirth) [inline lightly/loss/byol_loss.py:9]: Please remove `ExperimentParams` from `byol_loss.py`.

If there is the wish for better experiment tracking, open an issue about it and then we can discuss the next steps.

File: lightly/loss/byol_loss.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        self.layer = layer
+
+    def __str__(self):
+        return "Running experiment with the following hyper-parameters.\n" \
+               "Momentum: {}\nBeta: {}\nAugment Function: {}\nNet: {}\nProjection Size: {}\n" \
+               "Projection Hidden Size: {}\n" "Layer: {}".format(self.use_momentum, self.beta, self.augment_functions,
+                                                                 self.net, self.projection_size,
+                                                                 self.projection_hidden_size, self.layer)
+
+
```
Comment (philippmwirth) [inline lightly/loss/byol_loss.py:32]: This is a cool idea! We still have to find the best way to get the activations from intermediate levels of the backbone net. However, I don't think `byol_loss.py` is the right place for such a feature.

Please remove `NetWrapper` from this file and feel free to make a suggestion in #60 on how to solve this issue.

File: lightly/loss/byol_loss.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    @staticmethod
+    def default(val, def_val):
+        return val if val else def_val
+
+    @staticmethod
+    def flatten(t):
+        return t.reshape(t.shape[0], -1)
+
+
```
Comment (philippmwirth) [inline lightly/loss/byol_loss.py:94]: The classes `OnlineModel` and `TargetModel` will receive an abstract parent class with #57 (I will make a PR soon).

IMO it would be best to reuse the parent class as soon as it's ready.

File: lightly/loss/byol_loss.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            old_weight, up_weight = ma_params.data, current_params.data
+            ma_params.data = self._update_average(old_weight, up_weight)
+
+    def _update_average(self, old, new):
+        if self.ema:
+            return old * self.beta + (1 - self.beta) * new
+        else:
+            return new
+
+
```
Comment (philippmwirth) [inline lightly/loss/byol_loss.py:129]: The implementation of the loss should follow the style of the current implementations. I.e. the loss takes as input the outputs of the model. See:
- [Lightly at a Glance: Training](https://docs.lightly.ai/getting_started/lightly_at_a_glance.html)
- [NTXentLoss implementation](https://github.com/lightly-ai/lightly/blob/develop/lightly/loss/ntx_ent_loss.py)
- [Negative Cosine Similarity Loss by busycalibrating](https://github.com/lightly-ai/lightly/pull/61/files)

This allows for a more flexible setup where a single loss function can be combined with different architectures.

File: lightly/loss/byol_loss.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            target_encoder = self.target.encoder
+            target_proj_one = target_encoder(image_one).detach()
+            target_proj_two = target_encoder(image_two).detach()
+
+        loss_one = self._loss_fn(online_pred_one, target_proj_two.detach())
+        loss_two = self._loss_fn(online_pred_two, target_proj_one.detach())
+
+        loss = loss_one + loss_two
+        return loss.mean()
+
```
Comment (philippmwirth) [inline lightly/loss/byol_loss.py:177]: Please refrain from using static methods and use module level functions instead:

```python
def _neg_cosine_loss(x, y):
   """Calculates the asymmetric negative cosine loss.

   """
   pass

class BYOLLoss(nn.Module):
...

def forward(self, x):
   """Uses neg_cosine_loss to calculate the symmetric negative cosine loss.

   """
   pass
```

Source: [Google Python Styleguide](https://google.github.io/styleguide/pyguide.html#217-function-and-method-decorators)

File: lightly/loss/byol_loss.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    def _update_average(self, old, new):
+        if self.ema:
+            return old * self.beta + (1 - self.beta) * new
+        else:
+            return new
+
+
+class BYOLLoss:
+    """
```
Comment (philippmwirth) [inline lightly/loss/byol_loss.py:131]: For the docstrings format, please check out the [Google Python Styleguide](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings) to see how we do it :)

File: lightly/models/mlp.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
+from torch import nn
+
+
```
Comment (philippmwirth) [inline lightly/models/mlp.py:3]: #57 will also introduce utility classes for the projection heads and predictor heads.
